### PR TITLE
Respect types when evaluating unary and binary operations in `evalConstantExpr()`

### PR DESCRIPTION
**Pull request by blitz-1306**
## Changes
- [x] Respect types when evaluating unary and binary operations in `evalConstantExpr()`.
- [x] Add edge-cases to tests.

Regards.

### CODE REVIEW

File: src/types/infer.ts [hunk 1]
```diff
@@ -497,19 +497,6 @@ export class InferType {
         const a = this.typeOf(node.vLeftExpression);
         const b = this.typeOf(node.vRightExpression);
 
-        if (a instanceof NumericLiteralType && b instanceof NumericLiteralType) {
```
Comment (cd1m0) [inline src/types/infer.ts:500]: I'm worried about moving this line below. We want `InferType.typeOf(`4+3`)  to be `IntLiteral(7)`. I think moving this below would break this

**Review by cd1m0 — CHANGES_REQUESTED** [review]
I think moving the IntLiteral checks would break type inferrence (see comment below). I think we should come up  with another solution to avoid the inf recursion

File: src/types/infer.ts (hunk 1 above)
Comment (blitz-1306) [reply src/types/infer.ts]: I restored original logic in d4a7200 and found workaround in 2fcc3d9.

**Review by cd1m0 — APPROVED** [review]
LGTM
